File: libs/python/stonecharts/charts/vector_plot.py
"""Vector-plot chart renderer: ChartSpec -> SVG string.

Arrow glyphs on a numeric x/y plane. Each datum carries (x, y, direction, length).
Rides the shared cartesian frame with x_scale="linear" and include_zero=False.
"""

from __future__ import annotations

import math

from ..spec import Datum
from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, render_cartesian


def render_svg(spec) -> str:
    for s in spec.series:
        if not s.data_points:
            x_arr = s.x or [float(i) for i in range(len(s.data))]
            s.data_points = [Datum(x=x_arr[i], y=s.data[i]) for i in range(min(len(x_arr), len(s.data)))]
    return render_cartesian(spec, "Vector plot", "linear", _vector_marks, include_zero=False)


HEAD_LEN = 6.0
HEAD_ANGLE = 25.0


def _vector_marks(fr: CartesianFrame, p: list[str]) -> None:
    if fr.n <= 0:
        return

    K = len(fr.spec.series)

    vector_length = fr.spec.vector_length if fr.spec.vector_length else 20.0
    rotation_origin = fr.spec.rotation_origin if fr.spec.rotation_origin else "center"

    lmax = 0.0
    for s in fr.spec.series:
        ln = s.length or []
        for v in ln:
            if v > lmax:
                lmax = v

    def arrow_px(length: float) -> float:
        if lmax <= 0.0:
            return 0.0
        return vector_length * (length / lmax)

    ha = HEAD_ANGLE * math.pi / 180.0
    ca = math.cos(ha)
    sa = math.sin(ha)

    for si in range(K):
        s = fr.spec.series[si]
        st = fr.styles[si]

        x_arr = s.x or [float(i) for i in range(len(s.data))]
        y_arr = s.data
        dir_arr = s.direction or [0.0] * len(s.data)
        len_arr = s.length or [0.0] * len(s.data)

        n_pts = min(len(x_arr), len(y_arr), len(dir_arr), len(len_arr))

        stroke = st.fill
        line_width = s.line_width if s.line_width is not None else 1.5

        p.append(f'<g class="sc-series" data-series="{si}">')

        for i in range(n_pts):
            xv = x_arr[i]
            yv = y_arr[i]
            dv = dir_arr[i]
            lv = len_arr[i]

            cx = fr.xpix(xv)
            cy = fr.ypix(yv)

            rad = dv * math.pi / 180.0
            ux = math.sin(rad)
            uy = -math.cos(rad)

            big_l = arrow_px(lv)
            half = big_l / 2.0

            if rotation_origin == "start":
                ax, ay = cx, cy
            elif rotation_origin == "end":
                ax, ay = cx - ux * big_l, cy - uy * big_l
            else:
                ax, ay = cx - ux * half, cy - uy * half

            tailx, taily = ax, ay
            headx, heady = ax + ux * big_l, ay + uy * big_l

            lbx = headx + HEAD_LEN * ((-ux) * ca - (-uy) * sa)
            lby = heady + HEAD_LEN * ((-ux) * sa + (-uy) * ca)
            rbx = headx + HEAD_LEN * ((-ux) * ca + (-uy) * sa)
            rby = heady + HEAD_LEN * (ux * sa + (-uy) * ca)

            d = (
                f"M{tailx:.1f} {taily:.1f} L{headx:.1f} {heady:.1f} "
                f"M{lbx:.1f} {lby:.1f} L{headx:.1f} {heady:.1f} L{rbx:.1f} {rby:.1f}"
            )

            xlabel = esc(fmt_num(xv))
            ylabel = esc(fmt_num(yv))

            p.append(
                f'<path class="sc-vector sc-point" data-series="{si}" '
                f'data-series-name="{esc(s.name)}" '
                f'data-x="{xlabel}" data-y="{ylabel}" '
                f'data-direction="{esc(fmt_num(dv))}" data-length="{esc(fmt_num(lv))}" '
                f'data-color="{st.solid}" data-r="{fmt_num(line_width)}" data-r-hover="{fmt_num(line_width)}" '
                f'cx="{cx:.1f}" cy="{cy:.1f}" '
                f'd="{d}" fill="none" stroke="{stroke}" stroke-width="{fmt_num(line_width)}" '
                f'stroke-linecap="round" stroke-linejoin="round"/>'
            )

        p.append("</g>")

File: libs/go/vectorplot.go
package stonecharts

import (
	"fmt"
	"math"
	"strings"
)

func renderVectorPlotSVG(spec *ChartSpec) string {
	mod := *spec
	mod.Series = make([]Series, len(spec.Series))
	copy(mod.Series, spec.Series)

	for i := range mod.Series {
		s := &mod.Series[i]
		xArr := s.X
		if len(xArr) == 0 {
			xArr = make([]float64, len(s.Data))
			for j := range s.Data {
				xArr[j] = float64(j)
			}
		}
		n := len(xArr)
		if len(s.Data) < n {
			n = len(s.Data)
		}
		s.DataPoints = make([]Datum, n)
		for j := 0; j < n; j++ {
			s.DataPoints[j] = Datum{X: xArr[j], Y: s.Data[j]}
		}
	}
	return renderCartesian(&mod, "Vector plot", "linear", vectorPlotMarks, false)
}

const (
	vpHeadLen   = 6.0
	vpHeadAngle = 25.0
)

func vectorPlotMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}

	vectorLength := fdef(f.spec.VectorLength, 20)
	rotationOrigin := f.spec.RotationOrigin
	if rotationOrigin == "" {
		rotationOrigin = "center"
	}

	lmax := 0.0
	for _, s := range f.spec.Series {
		for _, v := range s.Length {
			if v > lmax {
				lmax = v
			}
		}
	}

	arrowPx := func(length float64) float64 {
		if lmax <= 0.0 {
			return 0.0
		}
		return vectorLength * (length / lmax)
	}

	ha := vpHeadAngle * math.Pi / 180.0
	ca := math.Cos(ha)
	sa := math.Sin(ha)

	for si, s := range f.spec.Series {
		st := f.styles[si]

		xArr := s.X
		if len(xArr) == 0 {
			xArr = make([]float64, len(s.Data))
			for i := range s.Data {
				xArr[i] = float64(i)
			}
		}
		yArr := s.Data
		dirArr := s.Direction
		if len(dirArr) == 0 {
			dirArr = make([]float64, len(s.Data))
		}
		lenArr := s.Length
		if len(lenArr) == 0 {
			lenArr = make([]float64, len(s.Data))
		}

		nPts := len(xArr)
		if len(yArr) < nPts {
			nPts = len(yArr)
		}
		if len(dirArr) < nPts {
			nPts = len(dirArr)
		}
		if len(lenArr) < nPts {
			nPts = len(lenArr)
		}

		stroke := st.fill
		lineWidth := 1.5
		if s.LineWidth > 0 {
			lineWidth = s.LineWidth
		}

		p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))

		for i := 0; i < nPts; i++ {
			xv := xArr[i]
			yv := yArr[i]
			dv := dirArr[i]
			lv := lenArr[i]

			cx := f.xpix(xv)
			cy := f.ypix(yv)

			rad := dv * math.Pi / 180.0
			ux := math.Sin(rad)
			uy := -math.Cos(rad)

			bigL := arrowPx(lv)
			half := bigL / 2.0

			var ax, ay float64
			switch rotationOrigin {
			case "start":
				ax, ay = cx, cy
			case "end":
				ax, ay = cx-ux*bigL, cy-uy*bigL
			default:
				ax, ay = cx-ux*half, cy-uy*half
			}

			tailx, taily := ax, ay
			headx, heady := ax+ux*bigL, ay+uy*bigL

			lbx := headx + vpHeadLen*((-ux)*ca-(-uy)*sa)
			lby := heady + vpHeadLen*((-ux)*sa+(-uy)*ca)
			rbx := headx + vpHeadLen*((-ux)*ca+(-uy)*sa)
			rby := heady + vpHeadLen*(-(-ux)*sa+(-uy)*ca)

			d := fmt.Sprintf("M%s %s L%s %s M%s %s L%s %s L%s %s",
				f1(tailx), f1(taily), f1(headx), f1(heady),
				f1(lbx), f1(lby), f1(headx), f1(heady), f1(rbx), f1(rby))

			p.WriteString(fmt.Sprintf(
				`<path class="sc-vector sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-direction="%s" data-length="%s" data-color="%s" data-r="%s" data-r-hover="%s" cx="%s" cy="%s" d="%s" fill="none" stroke="%s" stroke-width="%s" stroke-linecap="round" stroke-linejoin="round"/>`,
				si, esc(s.Name), esc(fmtNum(xv)), esc(fmtNum(yv)),
				esc(fmtNum(dv)), esc(fmtNum(lv)),
				st.solid, fmtNum(lineWidth), fmtNum(lineWidth),
				f1(cx), f1(cy), d, stroke, fmtNum(lineWidth)))
		}

		p.WriteString(`</g>`)
	}
}
